### PR TITLE
Java components for compile time injection

### PR DESCRIPTION
**Pull request by marcospereira**
# Status

This is ready to be reviewed.

* [x] Documentation (Hightlights and how to use the components)
* [x] Javadocs for all the components
* [x] Java version for `play.api.BuiltInComponentsFromContext`
* [x] Filters components like `play.api.NoHttpFiltersComponents` (maybe this should be done in a serie of smaller pull requests)
* [x] `*Components` for subprojects

## Purpose

Java components APIs similar to what we have in Scala for Compile Time Injection.

### CODE REVIEW

File: framework/src/play/src/main/java/play/BuiltInComponents.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+                messagesApi().asScala(),
+                langs().asScala(),
+                fileMimeTypes().asScala(),
+                httpConfiguration()
+        );
+    }
+
+    @Override
+    default Langs langs() {
+        play.api.Configuration configuration = new play.api.Configuration(configuration());
```
Comment (marcospereira) [inline framework/src/play/src/main/java/play/BuiltInComponents.java]: Add a private method to avoid repeatedly creating an `play.api.Configuration`.

**Comment by marcospereira** [comment]
This now depends of #7179 and #7180.

**Comment by wsargent** [comment]
Also see https://github.com/playframework/play-java-dagger2-example :-)

**Comment by marcospereira** [comment]
This is ready to be reviewed.

There is still some logger configuration aspects that can be improved (see documentation examples), but there is PR #7187 for that.

File: framework/src/play/src/main/java/play/BuiltInComponentsFromContext.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Override
+    public ApplicationLifecycle applicationLifecycle() {
+        return context.applicationLifecycle();
+    }
+
+    @Override
+    public Injector injector() {
+        // TODO do we need to register components like the scala version?
```
Comment (gmethvin) [inline framework/src/play/src/main/java/play/BuiltInComponentsFromContext.java]: Since this injector is used to build the application, it will still be a problem when we call `app.injector.instanceOf[Foo]` for the global state APIs. We could just choose not to have the injector be overridable at all and make this work only without global state, but it's not hard to get it working.

We probably should write a helper method to get the default injector instance that we call from both places.

Also, we need to remember to set/unset the ObjectMapper using Json.setObjectMapper. That will definitely still be in the API even if we recommend moving to something non-static for JSON.

File: documentation/manual/working/javaGuide/main/dependencyinjection/JavaCompileTimeDependencyInjection.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+Out of the box, Play provides a mechanism for runtime dependency injection - that is, dependency injection where dependencies aren't wired until runtime.  This approach has both advantages and disadvantages, the main advantages being around minimisation of boilerplate code, the main disadvantage being that the construction of the application is not validated at compile time.
+
+An alternative approach is to use compile time dependency injection.  At its simplest, compile time DI can be achieved by manually constructing and wiring dependencies.  Other more advanced techniques and tools exist, such as [Dagger](https://google.github.io/dagger/).  All of these can be easily implemented on top of constructors and manual wiring, so Play's support for compile time dependency injection is provided by providing public constructors and factory methods as API.
+
+> **Note**: If you're new to compile-time DI or DI in general, it's worth reading Adam Warski's [guide to DI in Scala](https://di-in-scala.github.io/) that discusses compile-time DI in general. While this is an explanation for Scala developers, it could also give you some insights about the advantages of Compile Time Injection.
```
Comment (marcospereira) [inline documentation/manual/working/javaGuide/main/dependencyinjection/JavaCompileTimeDependencyInjection.md]: Yeah. Not happy with this also. Will take a look at Dagger docs and see if they have something we can link here.

File: framework/src/play/src/main/java/play/BuiltInComponents.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+import play.libs.crypto.HMACSHA1CookieSigner;
+import play.mvc.FileMimeTypes;
+import scala.compat.java8.OptionConverters;
+
+import javax.inject.Provider;
+
+/**
+ * Helper to provide the Play built in components.
+ */
+public interface BuiltInComponents extends BaseComponents,
```
Comment (marcospereira) [inline framework/src/play/src/main/java/play/BuiltInComponents.java]: I though about that, but decide to keep like this because Scala API is just `play.api.BuiltInComponents` and I didn't what to move it.

File: framework/src/play/src/main/java/play/BuiltInComponentsFromContext.java (hunk 2 above)
Comment (marcospereira) [inline framework/src/play/src/main/java/play/BuiltInComponentsFromContext.java]: The trick point here is a circular dependency between some Java components (see `injector()` calls inside Java version of `BuiltInComponents`) and the injector itself (components that we need to add to have an injector on par with the Scala version).

To have a common injector initialization method, we would have to completely re-instantiate the Scala version of the components, but this shouldn't happen completely apart from the Java version (per instance, we have to honor `httpFilters` implementation) and we are then back to the circular dependencies here.

What I'm thinking is: if we are using dependency injection, we shouldn't touch the global state, right? Apparently this is current not true, for both compile-time and runtime di. For example, the following test:

```java
package play;

import org.junit.Test;
import play.components.BodyParserComponents;
import play.filters.components.HttpFiltersComponents;
import play.mvc.Http;
import play.mvc.Result;
import play.mvc.Results;
import play.routing.Router;
import play.routing.RoutingDsl;
import play.test.Helpers;

import static org.hamcrest.CoreMatchers.equalTo;
import static org.junit.Assert.assertThat;

public class BuiltInComponentsFromContextTest {

    class TestBuiltInComponentsFromContext extends BuiltInComponentsFromContext implements
            HttpFiltersComponents,
            BodyParserComponents {

        public TestBuiltInComponentsFromContext(ApplicationLoader.Context context) {
            super(context);
        }

        @Override
        public Router router() {
            return new RoutingDsl(defaultScalaParser(), javaContextComponents())
                    .GET("/").routeTo(() -> Results.ok("index"))
                    .build();
        }
    }

    @Test
    public void shouldProvideAApplication() {
        ApplicationLoader.Context context = ApplicationLoader.create(Environment.simple());
        Application application = new TestBuiltInComponentsFromContext(context).application();
        Helpers.running(application, () -> {
            Http.RequestBuilder request = Helpers.fakeRequest(Helpers.GET, "/");
            Result result = Helpers.route(application, request);
            assertThat(result.status(), equalTo(Helpers.OK));
        });
    }
}
```
It fails because we are still touching the global state:
```
[error] Test play.BuiltInComponentsFromContextTest.shouldProvideAApplication failed: java.lang.RuntimeException: java.lang.InstantiationException: play.api.http.HttpConfiguration, took 1.404 sec
[error]     ...
[error] Caused by: java.lang.InstantiationException: play.api.http.HttpConfiguration
[error]     ...
[error]     at play.api.mvc.SessionCookieBaker.COOKIE_NAME(Session.scala:75)
[error]     at play.api.mvc.SessionCookieBaker.COOKIE_NAME$(Session.scala:75)
[error]     at play.api.mvc.Session$.COOKIE_NAME(Session.scala:118)
```
Long story short: `Session.COOKIE_NAME` is global state that shouldn't be touched. Removing it will make the test pass even if are not adding components to the injector.

Comment (marcospereira) [inline framework/src/play/src/main/java/play/BuiltInComponentsFromContext.java]: Anyway, accessing global state inside injected components is out of the scope here. I will address it in another PR.

**Comment by gmethvin** [comment]
This seems like a great idea. The only question I have is how to deal with singletons, i.e. stateful objects that we want exactly one instance of. In scala the solution is to use a lazy val instead of a def. What's the answer for Java users? Should we provide some kind of helper? Are there any of the core components that should be made singleton?

**Comment by gmethvin** [comment]
If we clearly address the singleton problem this LGTM.

**Comment by marcospereira** [comment]
@gmethvin good question.

Since interfaces don't have state, we have two options:

1. Have a "cache" component that hold the singleton instances and make the interfaces use that "cache" to register and retrieve the singletons. I don't like this idea because it sounds complex and also because it kind of bring back the global state. Think for instance how to deal with this "cache" when we have more than one application running.

2. Delegate the singleton responsibility to `BuiltInComponentsFromContext` abstract class, instead of using default methods inside interfaces. So, the `BuiltInComponentsFromContext` will just handle these components as singletons for that specific `BuiltInComponentsFromContext` instance.

I more inclined to the second option, but still, users can decide to implement the interfaces without using `BuiltInComponentsFromContext` and then they will need to handle the singletons by themselves.

> Are there any of the core components that should be made singleton?

Maybe `application` since it wouldn't make sense to create more than one application per `BuiltInComponentsFromContext`.

WDYT?

**Comment by gmethvin** [comment]
I think it's fine to do it in `BuiltInComponentsFromContext`. Basically we need fields for all the things that should be singletons.

As far as which components should be singletons, we could take the safe approach and just do it for anything marked `@Singleton`. Otherwise we should go through one by one and make sure we know which ones actually need to be singletons. The main one I can think of is `ActorSystem` since the actor system is expensive to create and actually spins up thread pools. `Application` in theory shouldn't need it, but it would probably be a good idea just to make sure the application reference is always the same.

Anything in the `Context` should be fine since we already have only one instance there.

Basically anything that creates thread pools, is expensive to create, or holds state should be a singleton. There may be others I haven't thought of here.

**Comment by marcospereira** [comment]
@gmethvin take a look again, regarding the singleton behavior. The components that are annotated as singletons, but were not implemented here such as are:

1. `GzipFilter`: maybe it should not be a singleton, since its creation is pretty lightweight.
2. `play.filters.headers.SecurityHeadersFilter`: the same as gzip filter
3. `Formatters` & `FormFactory`: don't need to be singletons
4. BoneCP and HikariCP connection pools should be singletons, since its creations are expensive. But maybe we should just add sensible documentation here.

Anyway, this is ready to be reviewed again.

**Comment by gmethvin** [comment]
I think we should use a lazy initialization scheme, similar to what's described here: https://stackoverflow.com/a/29138089

The reason is that prevents us having to re-order the instantiations in the constructor if we change dependencies. (Though this isn't a problem for the end user as long as we get the order correct.)

We could expose a `lazily` helper so adding a custom singleton is as simple as this:
```java
private Supplier<PersonService> personService = lazily(() -> new PersonService(configuration()));
public PersonService personService() {
  return personService.get();
}
```

File: framework/src/play/src/main/java/play/BuiltInComponentsFromContext.java [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+    private final Supplier<CSRFTokenSigner> _csrfTokenSigner;
+    private final Supplier<Files.TemporaryFileCreator> _tempFileCreator;
+
+    private final Supplier<Injector> _injector;
+
+    public BuiltInComponentsFromContext(ApplicationLoader.Context context) {
+        // The order here matters
+        this.context = context;
+
+        this._injector = lazy(this::createInjector);
```
Comment (gmethvin) [inline framework/src/play/src/main/java/play/BuiltInComponentsFromContext.java]: Can't these be initialized above? Would be more readable.

File: framework/src/play/src/main/java/play/BuiltInComponentsFromContext.java [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+        this._csrfTokenSigner = lazy(this::createCsrfTokenSigner);
+
+        this._fileMimeTypes = lazy(this::createFileMimeTypes);
+        this._tempFileCreator = lazy(this::createTempFileCreator);
+        this._httpRequestHandler = lazy(this::createHttpRequestHandler);
+
+        this._application = lazy(this::createApplication);
+    }
+
+    private class LazySupplier<T> implements Supplier<T> {
```
Comment (gmethvin) [inline framework/src/play/src/main/java/play/BuiltInComponentsFromContext.java]: I think this should be a static class somewhere else. Maybe in `play.libs.F`? Then we could provide a static factory like `LazySupplier.lazy`

File: framework/src/play/src/main/java/play/BuiltInComponentsFromContext.java (hunk 5 above)
Comment (marcospereira) [inline framework/src/play/src/main/java/play/BuiltInComponentsFromContext.java]: You mean at field declaration?

File: framework/src/play/src/main/java/play/BuiltInComponentsFromContext.java (hunk 6 above)
Comment (marcospereira) [inline framework/src/play/src/main/java/play/BuiltInComponentsFromContext.java]: Ok.

File: framework/src/play/src/main/java/play/BuiltInComponentsFromContext.java (hunk 5 above)
Comment (gmethvin) [inline framework/src/play/src/main/java/play/BuiltInComponentsFromContext.java]: yes

**Comment by marcospereira** [comment]
See #7276.

File: framework/src/play/src/main/java/play/BuiltInComponentsFromContext.java [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+        ).asJava();
+    }
+
+    @Override
+    public Injector injector() {
+        return this._injector.get();
+    }
+
+    private Injector createInjector() {
+        // TODO do we need to register components like the scala version?
```
Comment (gmethvin) [inline framework/src/play/src/main/java/play/BuiltInComponentsFromContext.java]: @marcospereira Actually this is going to cause a bug because we use `injector()` in the `DefaultJavaHandlerComponents` to obtain instances. So we either need to add those things to the injector, or provide another way to provide action and body parser class instances. See https://github.com/playframework/playframework/blob/master/framework/src/play/src/main/scala/play/core/j/JavaAction.scala#L169
